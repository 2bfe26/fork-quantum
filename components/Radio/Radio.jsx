import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Colors from '../Colors';
import { Label } from '../shared';

const commonAttr = {
  borderWidth: '3px',
  height: '24px',
  width: '24px',
};

const Wrapper = styled.div`
  cursor: pointer;
  margin-right: 10px;
`;

const StyledLabel = styled(Label)`
  cursor: inherit;
  display: inline-block;

  &:before {
    content: ' ';
    border: 1px solid ${({ disabled }) => (disabled ? Colors.SECONDARY[100] : Colors.SECONDARY[300])};
    border-radius: 50%;
    display: inline-block;
    height: ${commonAttr.height};
    width: ${commonAttr.width};
    margin-left: -4px;
    margin-top: -4px;
    margin-right: 5px;

    ${({ checked }) => checked && `
      box-shadow: inset 0px 0px 0 3px white;
    `}
    cursor: inherit;
    position: relative;
    top: 6px;
  }

  &:hover:before {
    border: 1px solid ${({ disabled }) => (disabled ? Colors.SECONDARY[100] : Colors.PRIMARY[500])};
    box-shadow: inset 0 0 0 3px ${Colors.WHITE};
  }
`;

const StyledInput = styled.input`
  display: none;

  &:checked + ${StyledLabel}:before {
    background-color: ${Colors.PRIMARY[500]};
    box-shadow: inset 0 0 0 3px ${Colors.WHITE};
  }

  &:disabled:checked + ${StyledLabel}:before {
    background-color: ${Colors.SECONDARY[100]};
    box-shadow: inset 0 0 0 3px ${Colors.WHITE};
  }
`;

const Radio = ({
  id, label, onChange, disabled, ...rest
}) => (
  <Wrapper>
    <StyledInput
      {...rest}
      id={id}
      type="radio"
      disabled={disabled}
      onChange={e => onChange(e, { checked: e.target.value })}
    />
    <StyledLabel
      {...rest}
      disabled={disabled}
      htmlFor={id}
    >
      {label}
    </StyledLabel>
  </Wrapper>
);

Radio.displayName = 'Radio';

Radio.defaultProps = {
  onChange: () => { },
  disabled: false,
};

Radio.propTypes = {
  /** Id to associate with label */
  id: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  /** Label that will be displayed on browser */
  label: PropTypes.string.isRequired,
  /** On change event handle function */
  onChange: PropTypes.func,
};

export default Radio;
