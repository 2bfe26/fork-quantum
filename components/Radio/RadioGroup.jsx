import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Group = styled.div`
  align-items: flex-start;
  display: flex;
  flex-direction: column;

  span {
    margin-bottom: 20px;
  }

  ${props => props.inline && `
    align-items: unset;
    flex-direction: unset;
  `}
`;

/**
 * Group for Radio components.
 */
const RadioGroup = props => <Group {...props} />;

RadioGroup.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

export default RadioGroup;
