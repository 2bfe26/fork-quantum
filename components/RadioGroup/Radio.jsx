import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Colors from '../Colors';

const StyledLabel = styled.div.attrs(props => ({
  role: 'radio',
  tabIndex: -1,
  'aria-checked': 'false',
}))`
  cursor: pointer;
  display: inline-block;

  &:before {
    content: ' ';
    border-width: 1.5px;
    border-style: solid;
    border-color: ${Colors.BLACK[400]};
    ${props =>
      props.disabled &&
      `
      background-color: ${Colors.BLACK[200]};
    `}
    border-radius: 50%;
    display: inline-block;
    height: 16px;
    width: 16px;
    margin-right: 5px;

    position: relative;
    top: 4px;
  }

  &[aria-checked='true']:before {
    background-color: ${Colors.BLUE[500]};
    border-color: ${Colors.BLUE[500]};
    box-shadow: inset 0 0 0 3.5px ${Colors.WHITE};
  }

  &[aria-checked='true']:hover:before,
  &[aria-checked='true']:focus:before {
    box-shadow: inset 0 0 0 3.5px ${Colors.WHITE},
      0 2px 6px 0 ${Colors.BLUE[50]};
  }

  &:hover:before,
  &:focus:before {
    border-color: ${Colors.BLUE[500]};
    box-shadow: 0 2px 6px 0 ${Colors.BLUE[50]};
  }

  &[disabled] {
    color: ${Colors.BLACK[400]};
  }

  &[disabled]:before {
    background-color: ${Colors.BLACK[200]};
  }

  &[disabled]:hover {
    cursor: not-allowed;
  }

  &[disabled]:hover:before,
  &[disabled]:focus:before {
    border-color: ${Colors.BLACK[400]};
    box-shadow: none;
  }

  &[aria-checked='true']:disabled:before {
    background-color: ${Colors.BLACK[400]};
    border-color: ${Colors.BLACK[400]};
    box-shadow: inset 0 0 0 3.5px ${Colors.WHITE};
  }
`;

class Radio extends React.Component {
  constructor(props) {
    super(props);

    this.KEYS = Object.freeze({
      RETURN: 13,
      SPACE: 32,
      LEFT: 37,
      UP: 38,
      RIGHT: 39,
      DOWN: 40,
    });

    this.radioRef = React.createRef();
  }

  shouldComponentUpdate(nextProps) {
    if (nextProps['aria-checked']) {
      this.radioRef.focus();
    }

    return true;
  }

  onClick = () => {
    const { check, disabled } = this.props;

    if (disabled) return;

    check(this.props);
  };

  onKeyDown = event => {
    const { checkPrevious, checkNext, check } = this.props;
    let changed;

    switch (event.keyCode) {
      case this.KEYS.SPACE:
        check(this.props);
        changed = true;
        break;

      case this.KEYS.RETURN:
        check(this.props);
        changed = true;
        break;

      case this.KEYS.UP:
        checkPrevious(this.props);
        changed = true;
        break;

      case this.KEYS.DOWN:
        checkNext(this.props);
        changed = true;
        break;

      case this.KEYS.LEFT:
        checkPrevious(this.props);
        changed = true;
        break;

      case this.KEYS.RIGHT:
        checkNext(this.props);
        changed = true;
        break;

      default:
        break;
    }

    if (changed) {
      event.stopPropagation();
      event.preventDefault();
    }
  };

  render() {
    const { id, label, ...rest } = this.props;

    return (
      <StyledLabel
        {...rest}
        onClick={this.onClick}
        onKeyDown={this.onKeyDown}
        ref={radio => (this.radioRef = radio)}
      >
        {label}
      </StyledLabel>
    );
  }
}

Radio.displayName = 'Radio';

Radio.defaultProps = {
  disabled: false,
};

Radio.propTypes = {
  /** Id to associate with label */
  value: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  /** Label that will be displayed on browser */
  label: PropTypes.string.isRequired,
};

export default Radio;
