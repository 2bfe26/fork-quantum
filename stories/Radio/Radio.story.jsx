import React from 'react';
import { storiesOf } from '@storybook/react';
import styled from 'styled-components';
import {
  Heading,
  AutoPropsApi,
  HowToImport,
  CodeExample,
  TabbedView,
  Tab,
} from '@catho-private/quantum-storybook-ui';

import { Radio, RadioGroup } from '../../components/Radio';
import { Col, Row } from '../../components/Grid';

const RadioGroupWrapper = styled.div`
  padding: 15px;
`;

const StyledDiv = styled.div`
  margin-bottom: 40px;
`;

const RadioBlock = (
  <RadioGroup>
    <Radio id="radio" value="radio" name="radio" label="Click me!" />
    <Radio id="radioTwo" value="radioTwo" name="radio" label="Click me!" />
  </RadioGroup>
);

const RadioInline = (
  <RadioGroup inline>
    <Radio
      id="radioInline"
      value="radio"
      name="radioInline"
      label="Click me!"
    />
    <Radio
      id="radioTwoInline"
      value="radioTwo"
      name="radioInline"
      label="Click me!"
    />
  </RadioGroup>
);

const BoxedRadio = (
  <RadioGroup boxed inline>
    <Radio id="radioBoxed" value="radio" name="radioInline" label="Yes" />
    <Radio id="radioTwoBoxed" value="radioTwo" name="radioInline" label="No" />
  </RadioGroup>
);

const stories = storiesOf('Forms', module);
stories.add('Radio Button', () => (
  <Heading name="Radio" title="<RadioGroup />">
    <TabbedView>
      <Tab title="Usage">
        <HowToImport
          importModules="RadioGroup, Radio"
          componentName="RadioGroup"
        />

        <StyledDiv>
          <p>We provide two components to use Radio Buttons.</p>
          <p>
            <code>{'<RadioGroup>'}</code> and <code>{'<Radio>'}</code> must be
            used together and can be displayed inlined or blocked
          </p>
        </StyledDiv>

        <h4>RadioGroup with block display</h4>
        <Row>
          <Col phone={6}>
            <CodeExample component={RadioBlock} showTitle={false} />
          </Col>

          <Col phone={6}>
            <RadioGroupWrapper>{RadioBlock}</RadioGroupWrapper>
          </Col>
        </Row>

        <h4>RadioGroup with inline display</h4>
        <Row>
          <Col phone={6}>
            <CodeExample component={RadioInline} showTitle={false} />
          </Col>

          <Col phone={6}>
            <RadioGroupWrapper>{RadioInline}</RadioGroupWrapper>
          </Col>
        </Row>

        <h4>Boxed RadioGroup</h4>
        <Row>
          <Col phone={6}>
            <CodeExample component={BoxedRadio} showTitle={false} />
          </Col>

          <Col phone={6}>
            <RadioGroupWrapper>{BoxedRadio}</RadioGroupWrapper>
          </Col>
        </Row>
      </Tab>
      <Tab title="API">
        <AutoPropsApi component={Radio} title="Radio" />
        <AutoPropsApi component={RadioGroup} title="RadioGroup" />
      </Tab>
    </TabbedView>
  </Heading>
));
