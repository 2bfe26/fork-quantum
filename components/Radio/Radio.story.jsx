import React from 'react';
import { storiesOf } from '@storybook/react';
import RadioGroup from './RadioGroup';
import Radio from './Radio';

const stories = storiesOf('3. Forms', module);
stories
  .add('Radio Button', () => (
    <RadioGroup>
      <Radio id="radio" value="radio" name="radio" label="Click me!" />
      <Radio id="radioTwo" value="radioTwo" name="radio" label="Click me!" />
    </RadioGroup>
  ));
