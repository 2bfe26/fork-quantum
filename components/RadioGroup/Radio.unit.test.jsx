import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import Radio from './Radio';

describe('<RadioGroup.Radio />', () => {
  const checkMock = jest.fn();
  const checkPreviousMock = jest.fn();
  const checkNextMock = jest.fn();

  const KEYS = Object.freeze({
    SPACE: 32,
    LEFT: 37,
    UP: 38,
    RIGHT: 39,
    DOWN: 40,
  });

  const radioExample = (
    <Radio
      value="Hello"
      label="Hello"
      check={checkMock}
      checkNext={checkNextMock}
      checkPrevious={checkPreviousMock}
    />
  );

  const wrapper = mount(radioExample);
  const shalloWrapper = shallow(radioExample);

  it('Should match the snapshot', () => {
    const tree = renderer.create(radioExample).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('properties', () => {
    it('Should have defaults', () => {
      const {
        type: {
          attrs: [attributes],
        },
      } = shalloWrapper.get(0);
      expect(attributes).toHaveProperty('role', 'radio');
      expect(attributes).toHaveProperty('aria-checked', 'false');
      expect(attributes).toHaveProperty('tabIndex', -1);
    });
  });

  describe('events', () => {
    const eventDefault = {
      stopPropagation() {},
      preventDefault() {},
    };

    it('Should check when clicked and when pressing space bar', () => {
      const space = {
        ...eventDefault,
        keyCode: KEYS.SPACE,
      };

      wrapper.simulate('click');
      expect(checkMock).toBeCalledWith(wrapper.props());

      wrapper.simulate('keydown', space);
      expect(checkMock).toBeCalledWith(wrapper.props());
    });

    it('Should check next radio when arrow right or down is triggered', () => {
      const right = {
        ...eventDefault,
        keyCode: KEYS.RIGHT,
      };

      const down = {
        ...eventDefault,
        keyCode: KEYS.DOWN,
      };

      wrapper.simulate('keydown', right);
      expect(checkNextMock).toBeCalledWith(wrapper.props());

      wrapper.simulate('keydown', down);
      expect(checkNextMock).toBeCalledWith(wrapper.props());
    });

    it('Should check previous radio when arrow left or up is triggered', () => {
      const up = {
        ...eventDefault,
        keyCode: KEYS.UP,
      };

      const left = {
        ...eventDefault,
        keyCode: KEYS.LEFT,
      };

      wrapper.simulate('keydown', up);
      expect(checkPreviousMock).toBeCalledWith(wrapper.props());

      wrapper.simulate('keydown', left);
      expect(checkPreviousMock).toBeCalledWith(wrapper.props());
    });

    it('Should focus when aria-checked change', () => {
      const radio = wrapper.instance().radioRef;
      jest.spyOn(radio, 'focus');

      wrapper.setProps({ 'aria-checked': true });

      expect(radio.focus).toHaveBeenCalledTimes(1);
    });
  });
});
