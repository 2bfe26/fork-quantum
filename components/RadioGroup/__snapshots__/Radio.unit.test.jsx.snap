// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component Should match the snapshot 1`] = `
<div
  className="Radio__Wrapper-sc-1diyv3n-1 corTqy"
  name="radio"
  value="1"
>
  <input
    className="Radio__StyledInput-sc-1diyv3n-2 fTTWXf"
    disabled={false}
    id="radio"
    name="radio"
    onChange={[Function]}
    type="radio"
    value="1"
  />
  <label
    className="Label-sc-1ny5bqy-0 Radio__StyledLabel-sc-1diyv3n-0 fetNXm"
    disabled={false}
    htmlFor="radio"
    name="radio"
    value="1"
  >
    Hello
  </label>
</div>
`;
